src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocksHelper.cs

```diff
@@ -0,0 +1,361 @@
+                    throw new Exception("Connection failed");
+                int bytesToSkip = buffer[3] switch
+                {
+                    ATYP_IPV4 => 5,
+                    ATYP_IPV6 => 17,
+                    ATYP_DOMAIN_NAME => buffer[4] + 2,
+                    _ => throw new Exception("Unknown address type")
+                };
+                await ReadToFillAsync(stream, buffer.AsMemory(0, bytesToSkip), async).ConfigureAwait(false);
+                // response address not used
+            }
+            finally
+            {
+                ArrayPool<byte>.Shared.Return(buffer);
+            }
+        }
+
+        private static async ValueTask EstablishSocks4TunnelAsync(Stream stream, bool isVersion4a, string host, int port, Uri proxyUri, ICredentials? proxyCredentials, bool async, CancellationToken cancellationToken)
+        {
+            byte[] buffer = ArrayPool<byte>.Shared.Rent(BufferSize);
+
+            try
+            {
+                // https://www.openssh.com/txt/socks4.protocol
+
+                // +----+----+----+----+----+----+----+----+----+----+....+----+
+                // | VN | CD | DSTPORT |      DSTIP        | USERID       |NULL|
+                // +----+----+----+----+----+----+----+----+----+----+....+----+
+                //    1    1      2              4           variable       1
+                string? username = proxyCredentials?.GetCredential(proxyUri, "")?.UserName;
+                buffer[0] = ProtocolVersion4;
+                buffer[1] = CMD_CONNECT;
+
+                Debug.Assert(port > 0);
+                Debug.Assert(port < ushort.MaxValue);
+                buffer[2] = (byte)(port >> 8);
+                buffer[3] = (byte)port;
+
+                IPAddress? ipv4Address = null;
+                if (IPAddress.TryParse(host, out var hostIP))
+                {
+                    if (hostIP.AddressFamily == Sockets.AddressFamily.InterNetwork)
+                    {
+                        ipv4Address = hostIP;
+                    }
+                    else if (hostIP.IsIPv4MappedToIPv6)
+                    {
+                        ipv4Address = hostIP.MapToIPv4();
+                    }
+                    else
+                    {
+                        throw new Exception("SOCKS4 does not support IPv6.");
+                    }
+                }
+                else if (!isVersion4a)
+                {
+                    // SOCKS4 requires DNS resolution locally
+                    var addresses = async
+                        ? await Dns.GetHostAddressesAsync(host, Sockets.AddressFamily.InterNetwork, cancellationToken).ConfigureAwait(false)
+                        : Dns.GetHostAddresses(host);
+
+                    if (addresses.Length == 0)
+                    {
+                        throw new Exception("No suitable IPv4 address.");
+                    }
+
+                    ipv4Address = addresses[0];
```

Comment:
This could be IPv6.
We should loop over the returned addresses and select the first IPv4 one if available, or throw the same exception as above if not.

Reply:
Can it still return ipv6 address when AddressFamily passed for ipv4?

Reply:
I don't think so.

An easier sollution is then to pass the `AddressFamily` flag to the sync call too.